### ISSUE TITLE
[OpenTX V2.3.8] - Wrong signature warning on X12S when try to flash MPM

### ISSUE THREAD

**Issue by robustini** [opentx/opentx#7221]
Could you check if this is indeed the case?
On X10S Express and Jumper T16 no problem, when I try on X12S I have this error.

**Comment by 3djc** [comment]
Could you share the file you are trying to flash ? Fairly strange as the code should be the same

**Comment by robustini** [comment]
For the multimodule or the tx?
If you mean the MPM firmware I flash the same file in the others tx and work fine.

**Comment by robustini** [comment]
[multi-stm-1.3.0.52_EXT.zip](https://github.com/opentx/opentx/files/3989969/multi-stm-1.3.0.52_EXT.zip)

**Comment by 3djc** [comment]
For the multi, it will make it easier to replicate. Also is your x12 using internalaccess ?

**Comment by robustini** [comment]
Yes, internalaccess enabled.

**Comment by 3djc** [comment]
Ok i will work on it tomorrow if I can replicate on simu. If not, I’ll need to find someone with x12

**Comment by 3djc** [comment]
OK, I cannot reproduce, but the wrong signature message seems to indicate the radio does not like the info comming back from the multi module bootloader. Does flashing this very module on another radio work ?

**Comment by robustini** [comment]
Yes, with other radio work, with the same module.
I removed that pin from the ISRM module for Multi-Trainer but I don't think that's the problem.
![isrm_pin](https://user-images.githubusercontent.com/3864712/71305960-0c561700-23db-11ea-8054-d1f93425ac5b.jpg)

**Comment by 3djc** [comment]
@raphaelcoeffic your thoughs ?

**Comment by robustini** [comment]
If the X12S uses this pin to flash the MPM then I've cut out the possibility of flashing MPM with OpenTX, but on the X10S Express I did the same thing and the flashing works.
But it seems to me somewhat strange that this is so.

**Comment by robustini** [comment]
https://youtu.be/6mlW7paMfAA

**Comment by 3djc** [comment]
Yes confirm it is a issue while talking to the module in BL mode

**Comment by robustini** [comment]
Hardware or software issue?

**Comment by 3djc** [comment]
Don’t know yet, someone with the right hardware is going to have to check that

**Comment by gvt70** [comment]
Just updated to 2.3.4 and tried updating MPM with  
https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/releases/download/1.3.0.47/multi-stm-opentx-taer-inv-v1.3.0.47.bin

X12S: 
"Firmware update error"
NoSync

- same error with x10s on OpenTX 2.3.4
// same error with x9d+ on OpenTX 2.3.4 // worked fine. My fault.

**Comment by 3djc** [comment]
I'm afraid that just means your module does not have a bootloader. I just flash mine from a x10 with 2.3.4.

robustini is also succeding from X10 btw

**Comment by robustini** [comment]
> 
> 
> Just updated to 2.3.4 and tried updating MPM with
> https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/releases/download/1.3.0.47/multi-stm-opentx-taer-inv-v1.3.0.47.bin
> 
> X12S:
> "Firmware update error"
> NoSync
> 
>     * same error with x10s on OpenTX 2.3.4
> 
>     * same error with x9d+ on OpenTX 2.3.4

Thanks for your feedback.
Here 2.3.3 work fine on X10S Express and T16, no words.

**Comment by robustini** [comment]
@gvt70 have you the right bootloader on your module?
Have you already flashed this module with your tx in a previous version of OpenTX?

**Comment by gvt70** [comment]
> @gvt70 have you the right bootloader on your module?
> Have you already flashed this module with your tx in a previous version of OpenTX?

Yes, Flashed the same BIN when it came out a couple of weeks ago using X9D.
But, I have flashed it with a STK last week. Could that disable flashing with TX?

**Comment by 3djc** [comment]
You could have wiped the bootloader i guess

**Comment by gvt70** [comment]
Sorry guys. BIG brain farth over here.
Just tested the MPM in a X9D again and it flashed fine.
X12S still no luck. Error as above. 
I will change above text about X9D+ to not cause any problems.

**Comment by 3djc** [comment]
Do you get Sync error or wrong signature ?

**Comment by gvt70** [comment]
Just tested with X10S also. Same error as on X12S : Sync Error
So only X9D flashes fine

**Comment by 3djc** [comment]
x10 works fine here

**Comment by robustini** [comment]
X10S Express work fine here, so stranger.
But X12S no, so there's a issue... :-)

**Comment by 3djc** [comment]
Yes, but your X12 has a VERY different issue than NoSync

**Comment by robustini** [comment]
> 
> 
> Yes, but your X12 has a VERY different issue than NoSync

No, see my video again, NoSync because i disabled the module in the second attempt... :P
Wrong Signature is the problem with the MPM enabled.

**Comment by 3djc** [comment]
Exactly what I'm saying, isn't it ?

**Comment by robustini** [comment]
@gvt70, you have NoSync on X12S with the MPM enabled in the Module section?

**Comment by gvt70** [comment]
I renamed the file to "multi-stm-v13047.bin" tried again. No luck. Tried a second time and it flashed succesfull with X10S. Tried flashing a couple of times and all went well.
X12S, same problem. First time it failed. Second time it flashed succesfull! Tried again and succesfull.

**Comment by robustini** [comment]
> 
> 
> I renamed the file to "multi-stm-v13047.bin" tried again. No luck. Tried a second time and it flashed succesfull with X10S. Tried flashing a couple of times and all went well.
> X12S, same problem. First time it failed. Second time it flashed succesfull! Tried again and succesfull.

Ok, I try to insist on the X12S too, let's see what happens.

**Comment by 3djc** [comment]
The "Wrong Signature" message that you have means you have passed the 'noSync' stage, and that you are talking to the module bootloader, but the radio doesn't like/understand what it is receiving. NoSync and Wrong Signature are not at all similar

**Comment by robustini** [comment]
No, it never works here, both MPM enabled and disabled, just different error, "Wrong Signature" when enabled and "NoSync" when disabled.
But I have ISRM on the X12S...

**Comment by robustini** [comment]
Nothing, always "Wrong Signature".
I start to hate this signature check! :P

**Comment by robustini** [comment]
> X12S, same problem. First time it failed. Second time it flashed succesfull! Tried again and succesfull.

Excellent, but he must always flash it, even on the first try.

**Comment by gvt70** [comment]
Btw, the X12S and X10S still have the standard IJXT module. Not the upgrade module.
Maybe you could try with the standard X12S module.

**Comment by robustini** [comment]
> 
> 
> Btw, the X12S and X10S still have the standard IJXT module. Not the upgrade module.
> Maybe you could try with the standard X12S module.

I've ISRM on X10S Express and work fine! :-)

**Comment by robustini** [comment]
I tried today on a friend's X12S, on the first attempt with MPM enabled he gave me "No Sync", on the second attempt he wrote it.
There is therefore a problem introduced recently, with 2.3.2 he always flash it on the first try.

**Comment by 3djc** [comment]
@robustini what exact type of external multi are you using ?

**Comment by robustini** [comment]
> @robustini what exact type of external multi are you using ?

I have tried with three different ones, it does the same thing, IRangeX IRX4, Jumper and URUAV TMX5, all flashed with the right bootloader, therefore flashable for example with Jumper T16 and X10S Express.

**Comment by markran** [comment]
I believe I'm having the same issue as @robustini. I have a IRangeX IRX4 and tried to flash the MPM firmware from OpenTX on my Horus x12s which has the original internal TX, not ISRM but got a "Firmware Update Error", "Wrong signature". I have bootloader 2.3.4 installed and "multi-stm-opentx-taer-inv-v1.3.0.47.bin" is on the MPM. 

I've tried doing it several times consecutively and also rebooting in between with no success. Any ideas?

**Comment by robustini** [comment]
The solution would be simple, in OpenTX a warning of "signature wrong, flash it anyway?", with the possibility of skipping the signature check.
Or find out where the bug is...
@markran, have you disabled the internal IXJT before flashing MPM?

**Comment by 3djc** [comment]
Again, no, signature check as nothing to do the firmware file, the signature is with the module, dialog with the module bootloader cannot be established, that cannot be bypassed

**Comment by robustini** [comment]
Then the thing is even more critical.
Could it work randomly?
What was introduced in OpenTX to block communication?

**Comment by 3djc** [comment]
well you know if it work randonly more than us if it is random or not. From what you have been saying it looks like it nevers works so not random.

Nothing is introduced to block communication. To update, the module is started in bootloader mode, and your radio is than unable to talk to it, hence is (very not descriptive) error message

**Comment by robustini** [comment]
If there is no possibility to log this event it will be difficult to understand where the problem lies.

**Comment by markran** [comment]
@robustini Yes, I tried it both with and without the internal IJXT enabled and got the same error message.

@3djc said
> ... dialog with the module bootloader cannot be established

I'd like to help figure this out but I don't really understand how it's supposed to work. So a few questions if you don't mind.

1. In what ways is the dialog with the module bootloader different than the dialog with the module for normal control? For instance, OpenTX appears to communicate with the firmware on the module when it shows the firmware version number in the model screen. Maybe there's a clue in some difference in the handshake between the two instances? In both instances (bootloader and normal control) the X12S and MPM are communicating via the same electrical pathway. So then is the issue most likely a difference in the software on one side or the other or are there other places the problem could be? 

2. Do we know (or suspect) if the issue is with the particular X12S or is the issue with the particular MPM module? This MPM module was successfully flashed via FTDI so the flashing app was able to establish dialog with the module bootloader. Would it help if someone having this error can resolder the FTDI connections to the MPM and confirm that the FTDI can establish communication with the bootloader? @robustini Have you tried reflashing your MPM via FTDI?

3. I also have an original Taranis here as well as an X-Lite Pro. Would it help to try flashing this MPM from one of those (assuming I rig up some wiring to the X-Lite Pro)? Just trying to think of ways to narrow down the issue.

**Comment by kilrah** [comment]
If you are sure your module has the bootloader version that allows flashing from the radio, and it works fully (control and telemetry/status) then it's not the module...

**Comment by robustini** [comment]
> @robustini Have you tried reflashing your MPM via FTDI?

I can flash the same MPM on the T16 and X10S Express without any problems, so the bootloader is correct.
It's a problem related to X12S and OpenTX, not on MPM.

**Comment by robustini** [comment]
No news about that?

**Comment by robustini** [comment]
> @robustini Have you tried reflashing your MPM via FTDI?

I can flash this MPM with USB without issue, inserted in the slot of the X12S turned off, and obviously with the tx on it works perfectly!
It is not a problem of the MPM, I flash the same module from the tx menu in the X10S Express and T16, so the problem is with X12S and ISRM.
I have many types of MPM and they all give the same flashing problem with X12S.
All friends who have the X12S with the original IXJT module can flash MPM from the tx without issue.

**Comment by kilrah** [comment]
Seems it always works if the external module is set to OFF in the selected model, just not if it's set to MULTI.

**Comment by robustini** [comment]
> 
> 
> Seems it always works if the external module is set to OFF in the selected model, just not if it's set to MULTI.

It would be nice if it were so, unfortunately it doesn't work anyway.
OpenTX V2.3.8 (latest master of today), this MPM flashes smoothly with X10S Express and T16.
Same problem with the external T16 MPM module istalled on this X12S with ISRM.
The problem is this tx with the ISRM installed, not the MPM.

https://youtu.be/PQVdgLLxBKw

**Comment by robustini** [comment]
> 
> 
> Seems it always works if the external module is set to OFF in the selected model, just not if it's set to MULTI.

It would be nice if it were so, unfortunately it doesn't work anyway.
OpenTX V2.3.8 DEV (latest master of today), this MPM flashes smoothly with X10S Express and T16.
Same problem with the external T16 MPM module istalled on this X12S with ISRM.
The LEDs of the MPM are clearly seen flashing in the video, so the TX tries to communicate but cannot.
The problem is this TX with the ISRM installed, not the MPM.

https://youtu.be/PQVdgLLxBKw

**Comment by kilrah** [comment]
My X12S has ISRM installed and external multi flashing always works if the external module is OFF.

**Comment by robustini** [comment]
@kilrah tranks for your feedback, for me it is law!
So do you think it could be that pin that I removed from the ISRM in order to use the multi trainer?

**Comment by robustini** [comment]
@kilrah thanks for your feedback, for me it is law!
So do you think it could be that pin that I removed from the ISRM in order to use the multi trainer?

**Comment by robustini** [comment]
Obviously this thing is relatively castrating, I prefer to have the multi trainer usable, so I will flash the MPM via USB in a moment.
It was only to understand if it is a problem caused by the absence of that pin, or if there is something else on my X12S.

**Comment by kilrah** [comment]
Since it doesn't work when multi is on there is certainly something that's wrong. Kinda low priority though.

**Comment by rmd4u** [comment]
Hello, I have the same problem with Horus X12S.

It is opentx 2.3.7 on Horus and I use the multimodul IRangeX IRX4.  The internal module is switched off and the external one is activated as Multi. 
After selecting flash file multi-stm-opentx-taer-inv-v1.3.0.76 or multi-stm-opentx-taer-inv-v1.3.0.91 the message "device reset" appears, then it makes a beep and the error message " wrong device" appears on the display. The same files with the same multimodule work on the T16 without problems.

**Comment by robustini** [comment]
> 
> 
> Hello, I have the same problem with Horus X12S.
> 
> It is opentx 2.3.7 on Horus and I use the multimodul IRangeX IRX4. The internal module is switched off and the external one is activated as Multi.
> After selecting flash file multi-stm-opentx-taer-inv-v1.3.0.76 or multi-stm-opentx-taer-inv-v1.3.0.91 the message "device reset" appears, then it makes a beep and the error message " wrong device" appears on the display. The same files with the same multimodule work on the T16 without problems.

Do you also have the ISRM module installed?

**Comment by rmd4u** [comment]
> Do you also have the ISRM module installed?

I have no ISRM module, only an original iXJT

**Comment by robustini** [comment]
> 
> 
> > Do you also have the ISRM module installed?
> 
> I have no ISRM module, only an original iXJT

Then it's strange, because with the IXJT you should be able to write it without problems.
However there is something strange in the X12S, on the other tx I don't have this problem.

**Comment by jornkaa** [comment]
Hi  - I have the same problem with an X10s - running openTx2.3.7.
Have been strugling a lot - did a flash of the module, but seemed to "mess" up. It is not broken, cuz it appears in the radio, but I am not able to bring it up in FlashMulti.
It will be detected in windows device manager as Maple, though..  and I am still able to select it on the radio.   ( have tried both with internal module both on and off.)
But I still get NoSync during upgrade.

Any advice would be greatly appreciated :-)

**Comment by markran** [comment]
I just tried this again on my Horus X12S with 2.3.7

> Seems it always works if the external module is set to OFF in the selected model, just not if it's set to MULTI.

I tried turning off the external module and it didn't work (fails whether the XJT module is on or off). "No Sync" error.

**Comment by robustini** [comment]
> 
> 
> I just tried this again on my Horus X12S with 2.3.7
> 
> > Seems it always works if the external module is set to OFF in the selected model, just not if it's set to MULTI.
> 
> I tried turning off the external module and it didn't work (fails whether the XJT module is on or off). "No Sync" error.

Thanks for your feedback.
This thing starts to be a bit random.
Before it seemed that it happened only with the ISRM, now also with the IXJT.
I tried again now, on the first attempt I've a "Wrong Signature", and on the next always "No Sync".

**Comment by brainbubblersbest** [comment]
I tried today flash my r xsr receiver in the field, via the sport connection in jr module bay.
got same message.
Cannot flash r9m or any receiver. tried it with external module set to off, and also with internal and external to off.
Internal Module flashing works.
Can anyone try flash a receiver this way to countercheck this?
Flashed now all my receivers with FRSKY USBtoSPORT Dongle, also the R9M.
STOCK Horus x12s with opentx 2.3.7 and options flex,multi,noheli and crossfire.

I came from 2.3.2 and i noticed that now when i execute a firmwarefile for external module, there is no switching on of the module supplypin?

**Comment by robustini** [comment]
> 
> 
> I tried today flash my r xsr receiver in the field, via the sport connection in jr module bay.
> got same message.
> Cannot flash r9m or any receiver. tried it with external module set to off, and also with internal and external to off.
> Internal Module flashing works.
> Can anyone try flash a receiver this way to countercheck this?
> Flashed now all my receivers with FRSKY USBtoSPORT Dongle, also the R9M.
> STOCK Horus x12s with opentx 2.3.7 and options flex,multi,noheli and crossfire.
> 
> I came from 2.3.2 and i noticed that now when i execute a firmwarefile for external module, there is no switching on of the module supplypin?

I tried now, on the first two attempts he didn't recognize an X8R, on the third attempt he flashed it without problems, I tried again three more times without turning off the tx and he always flashed it, it is this random thing that sent me in paranoia.
The question therefore is not only of the external modules, but also of the receivers.
Afterwards I try to flash my R9M 2019, i think it will work, the problem here is with MPM.

**Comment by robustini** [comment]
I tried again now, with MPM module off I always have "No Sync", with module on I've "Wrong Signature" instead.

**Comment by robustini** [comment]
I tried again now, with MPM module off I always have "No Sync", with module on I've "Wrong Signature" instead.
Since the signature check was introduced I've never been able to flash it on the X12S again with the ISRM installed.

**Comment by brainbubblersbest** [comment]
Sorry r9m will work for me. Have remind me that i have just wrong Firmware
with me, out at the field. But receivers dont.

Marco Robustini <notifications@github.com> schrieb am Mo., 18. Mai 2020,
10:05:

**Comment by robustini** [comment]
> Sorry r9m will work for me. Have remind me that i have just wrong Firmware with me, out at the field. But receivers dont.

Which receiver can't you flash, the R9?
It may seem absurd but I have tried now to flash an X8R with ACCST V2.1.0 firmware three times in succession and it has always been written.
So if my X12S manages to put the rx in bootloader mode and to always write them I think we have to check the code relating to the dialogue with the MPM to put it in the bootloader, because if the same MPM module on other tx is flashed and if on the X12S i can flash rx hassle free means there is something wrong with the code imho.

**Comment by brainbubblersbest** [comment]
Cant Flash the receivers via Sport. Without Module in bay. But it worked
with the old otx

Marco Robustini <notifications@github.com> schrieb am Di., 26. Mai 2020,
08:30:

**Comment by robustini** [comment]
And this is very strange because to me instead the flash of the rx works without any problem, only with MPM I can't.
I compile and use the latest 2.3.8 firmware from branch 2.3.

**Comment by robustini** [comment]
I discovered an interesting thing, which I hope will help developers.
I was trying the new "[MULTI-BootReloader](https://www.rcgroups.com/forums/showpost.php?p=44649589&postcount=17944)" procedure, which allows you to overwrite some incorrect bootloaders on some MPM modules.
Before carrying out this procedure I installed the drivers in the "Flash-Multi" package for the TMX5.
After the procedure were recognized by Windows as "Maple DFU", and then I correctly flashed them with "Flash-Multi".
I tried the same thing with a Jumper MPM, but on this the drivers are not installed because it is not in DFU mode but Serial.
So I said to myself, "I bet this is flashed regularly on the X12S", and in fact it is so!
So my X12S refuses to flash the MPM module that are in DFU, while it flashes the Jumper MPM without any problem because it is in Serial mode!
I hope this thing is therefore useful to intercept the reason why i cannot flash the "DFU MPM" on my X12S, which instead works on other TXs anyway, but it flashes the MPM Jumper without any problem.

![jumper_mpm_x12s](https://user-images.githubusercontent.com/3864712/83348272-45e86200-a32b-11ea-86af-b29a25b2433a.jpg)

**Comment by benlye** [comment]
@robustini - do any of the modules which you can't flash via the X12S have a physical bind button?

If so you could try holding the bind button down while you start the flash from the radio.  The radio power-cycles the module at the start of the flash process, and if the bind button is pressed when the module starts up it is supposed to stay in the bootloader forever.

I don't know if this will help, but it's worth a try.

**Comment by robustini** [comment]
Unfortunately there is no bind button installed on any of the MPMs I own.
Regardless it should work anyway, if with the Jumper MPM it works it should also be able to work with others.

**Comment by benlye** [comment]
That's a shame. It would have been an interesting diagnostic test. If anyone else has a module with a bind button it would be useful information. 

Could you try to make a video of the module LEDs when you try to flash it from the radio?

I suspect that maybe the module is leaving the bootloader before the flash starts, or maybe isn't even ready when the radio tries to sync.

The MULTI bootloader is pretty simple, really. If it sees the radio trying to flash it when it starts up it should stay in the bootloader and receive the firmware data. Either the radio isn't sending the data at the time the module expects it, or the radio is sending it but the module doesn't receive it. The window for this to happen is pretty small. 

I can't work out yet why a native USB module would be different to a Jumper JP4in1. The bootloader code is the same and isn't aware of the USB port.

**Comment by robustini** [comment]
Here the video: https://youtu.be/jSBcWzMZdh0

Procedure done with MPM disabled, "No Sync" message.
If I do it by activating it, it gives me "Wrong Signature".

**Comment by robustini** [comment]
@benlye @3djc I discovered another thing perhaps even more interesting, and I don't think it's normal.
With Flash-Multi if I go to read the modules "TMX5" via USB gives me the error "Firmware Signature Not Found" of the first image, if I read instead that of the Jumper MPM it works... why?
But all MPMs are detected functionally by the tx without problems.
DFU MPM do not allow correctly flash reading via USB?
All MPMs have been updated to version 1.3.1.5.

![image1](https://user-images.githubusercontent.com/3864712/83384675-0af83400-a3e8-11ea-920b-627608bf7562.png)

![image2](https://user-images.githubusercontent.com/3864712/83382882-56a8de80-a3e4-11ea-86fc-9fc8e55598a4.png)

**Comment by benlye** [comment]
Flash Multi definitely can read from an iRangeX module in DFU mode - that's what I used for almost all my testing and development.  Something very strange is going on there.

Can you do these things:
1. Post the verbose output of a read and write with your IRX4
2. Use Flash Multi to save the backup from the IRX4 and send it to me
3. If you have one, connect the IRX4 with a USB-to-serial adapter and try reading it, then send me the backup file.  It will also be interesting to know if Flash Multi is able to get the details when reading that way.

For the USB-to-serial backup, Flash Multi does a couple of things when it saves the file for you so I'd like the full backup it makes behind the scenes.  If you look in the %TEMP% folder and sort it by newest first you should see a file with a random name like 'tmpCF39.tmp' with a size of 128KB.  Can you send that file to me?

Last thing, can you check that you are using the libusbK driver for the DFU device?
https://github.com/benlye/flash-multi/blob/master/doc/Troubleshooting.md#re-installing-the-maple-dfu-device-drivers

Feel free to email me directly, my email address is in my Github profile.

**Comment by robustini** [comment]
@benlye , you have mail.
Maple Driver are already updated.
Here the verbose output when flash this TMX5 MPM.
Immediately after flashing it Windows recognizes it as a COM "Multi-Module USB Serial", disconnecting and reconnecting it finds it as Maple.

![immagine](https://user-images.githubusercontent.com/3864712/83385620-dc7b5880-a3e9-11ea-98f9-6b80392a4d31.png)

**Comment by robustini** [comment]
I've done other tests, I'll update you.
After flashing the IRX4 module with the X10S Express, now I can write it with the X12S as well, I have no words!
I did the same thing with the TMX5 but the problem remains, I can't flash it with the X12S.
So I'm starting to think it's a problem with this MPM, but that I can still flash on the X10S Express, so I don't understand... :p
I soldered the USB port on the IR4X on the fly and it works in DFU reading with Flash-Multi.
So the problem is related to TMX5.

**Comment by 3djc** [comment]
There is a possibility that signature check is correctly implemented in x12s and for some reason not in x10e

**Comment by robustini** [comment]
> 
> 
> There is a possibility that signature check is correctly implemented in x12s and for some reason not in x10e

Even the T16 writes all my MPM without problems.
I think instead at this point it is the TMX5 which is badly made.

**Comment by robustini** [comment]
Damn... :-)
TMX5 MPU.... :-(

![20200601_101028-0](https://user-images.githubusercontent.com/3864712/83389794-0421ef00-a3f1-11ea-9828-863b89a0d99d.jpg)

**Comment by benlye** [comment]
C8T6 = 64KB flash.  It should be a CBT6, with 128KB flash.

![en STM32F103_line_LN1565](https://user-images.githubusercontent.com/23322579/83391119-e2be0480-a3ea-11ea-9360-d35dd28b6a38.jpg)

Oh well, at least we know the problem now ;-)

You might want to try to open a case with the vendor and/or reseller - they may not even know that this happened.

Honestly, I'm surprised it ever worked.

**Comment by robustini** [comment]
Thanks @benlye to point me in the right direction.
So don't buy [this module](https://www.banggood.com/URUAV-TMX5-4-IN-1-A7105-CC2500-CYRF6936-NRF24L01+-Multi-Protocol-Transmitter-Module-Compatible-FrSky-X9D-X10-Transmitter-p-1525898.html?rmmds=search&cur_warehouse=CZ).
I hope @pascallanger put it on the black list.

**Comment by kilrah** [comment]
This has nothing to do with the flashing issues on X12S that still exist.

**Comment by robustini** [comment]
> 
> 
> This has nothing to do with the flashing issues on X12S that still exist.

I agree, I replaced on the TMX5 the STM32 with 128 Kb of flash, taken from an old Naze32, it works perfectly but continues not to be flashed by the X12S.

![5](https://user-images.githubusercontent.com/3864712/83406929-02b3ef00-a410-11ea-8206-b3cd3cf9635a.jpg)

**Comment by rmd4u** [comment]
> 
> 
> @benlye , you have mail.
> Maple Driver are already updated.
> Here the verbose output when flash this TMX5 MPM.
> Immediately after flashing it Windows recognizes it as a COM "Multi-Module USB Serial", disconnecting and reconnecting it finds it as Maple.
> 
> ![immagine](https://user-images.githubusercontent.com/3864712/83385620-dc7b5880-a3e9-11ea-98f9-6b80392a4d31.png)

Hello,
where can i find the version 1.3.1.5?

**Comment by kilrah** [comment]
You build it yourself
https://www.multi-module.org/advanced-topics/compiling-the-firmware

**Comment by rmd4u** [comment]
> You build it yourself
> https://www.multi-module.org/advanced-topics/compiling-the-firmware

i have flashed the 1.3.1.1 over FTDI, the bootloader is transferred with it.  There was no change in the behaviour of the Horus X12s. Still flashing via TX does not work afterwards. The self-compiling is puzzling. I don't know the files which I have to use in the current [DIY-Multiprotocol-TX-Module-master.zip](https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/archive/master.zip).
Installed is Arduino 1.8.50 and my module is an IRangeX without USB connection.

![grafik](https://user-images.githubusercontent.com/64751192/83619572-58ada180-a58c-11ea-83e3-97b6adcb7c99.png)

**Comment by robustini** [comment]
> 
> 
> > You build it yourself
> > https://www.multi-module.org/advanced-topics/compiling-the-firmware
> 
> i have flashed the 1.3.1.1 over FTDI, the bootloader is transferred with it. There was no change in the behaviour of the Horus X12s. Still flashing via TX does not work afterwards. The self-compiling is puzzling. I don't know the files which I have to use in the current [DIY-Multiprotocol-TX-Module-master.zip](https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/archive/master.zip).
> Installed is Arduino 1.8.50 and my module is an IRangeX without USB connection.
> 
> ![grafik](https://user-images.githubusercontent.com/64751192/83619572-58ada180-a58c-11ea-83e3-97b6adcb7c99.png)

Did you disable the module in the tx parameters before flashing?

**Comment by robustini** [comment]
Try with this one but turn off the module before flashing.
Now I can properly flash both IRangeX IRX4 and Jumper MPM but not TMX5, this one I flash correctly in the X10S Express, so the X12S definitely has something different on the hardware level, or there is a problem in the code.

[multi-stm-1.3.1.8.zip](https://github.com/opentx/opentx/files/4722408/multi-stm-1.3.1.8.zip)

**Comment by rmd4u** [comment]
> Try with this one but turn off the module before flashing.
> Now I can properly flash both IRangeX IRX4 and Jumper MPM but not TMX5, this one I flash correctly in the X10S Express, so the X12S definitely has something different on the hardware level, or there is a problem in the code.

Thanks for the Bootloader.
first I read out the module with multi flash3.0
![grafik](https://user-images.githubusercontent.com/64751192/83654234-181a4c00-a5bd-11ea-9a35-46a3dbb8ac73.png)
then I flashed the bootloader and read it again
![grafik](https://user-images.githubusercontent.com/64751192/83654378-50ba2580-a5bd-11ea-843b-7627e91d1f3c.png)
Unfortunately it is still not possible to flash the multifile with the X12

**Comment by robustini** [comment]
> Unfortunately it is still not possible to flash the multifile with the X12

Which MPM do you have?

**Comment by rmd4u** [comment]
> Which MPM do you have?

I have an IRangeX without USB connection

**Comment by robustini** [comment]
> I have an IRangeX without USB connection

Very strange, I have the same MPM and I can flash it, but my X12S installs the ISRM module with the pin removed to be able to operate both modules simultaneously, maybe this makes the difference.

**Comment by kkriegel** [comment]
Hi Folks,
Hopefully this post will help. I spent today upgrade testing 3 different IrangeX multi modules on 4 different transmitters. 

**In my case, it was only a hardware revision 1.3 Irx4+ which worked at all. It worked across all transmitters tested. Based on this outcome, I do not believe this is specific to the X12s.**

In summary, only my Irx4+ with hardware rev 1.3 upgrades from the Tx, and only after updating the boot loader per Ben Lye’s excellent instructions here:
https://www.rcgroups.com/forums/showpost.php?p=44649589&postcount=17944
Once the boot loader was updated, the Irx4+ upgraded in all transmitters without issue. I noticed that I needed to create a model that powered off the multi module in order for the upgrade to work.

The only error observed throughout my testing was the SYNC error.

**Transmitters tested (all running opentx 2.3.7 pre-built binaries):**
- Taranis X9D+, sticker in module bay says "A01 US PRC 18/04/2016"
- Taranis QX7, hardware revision unknown
- Horus X12S, sticker in module bay says "A01 US AH"
- Horus X12S, sticker in module bay says "A01 US RE"

**Modules tested:**
- 1 IrangeX Irx4, HW revision 0.8
- 1 IrangeX Irx4+, HW revision 1.2
- 1 IrangeX Irx4+, HW revision 1.3, purchased 2020-05-16


Thanks for all of your excellent work on opentx.

-Ken

**Comment by robustini** [comment]
Thanks @kkriegel for your feedback.
The problem of the difficulty of flashing is often also attributable to the type of MPM combined with some TXs.
But the X12S has something different hardware or software that differentiates it from other tx.
In fact some MPMs on this TX cannot be flashed even by turning the module off the module from the model settings, when instead the same MPM flashes without problems on other tx, so it's not just a question of bootloader, which is always better to update.
With IrangeX IRX4 I've no problem, with the TMX5 yes, also by skipping the protection diode on the communication pin, but I can instead flash it for example on X10S Express.
So the X12S has something different.
I personally believe that the MPM Jumper JP4IN1-SE is the best, as I hope the JP5IN1 will be on the way.

**Comment by kkriegel** [comment]
@robustini:
I have updated my comment with a conclusion, bolded. I had deliberately not opined on the root cause, allowing folks to reach their own conclusions. I will be very explicit here: Based on the experimental outcome, the X12S exhibited no difference in behavior from other transmitters running OpenTx 2.3.7. It was only the latest Irx4 hardware that solved the problem.

**Comment by robustini** [comment]
> 
> 
> @robustini:
> I have updated my comment with a conclusion, bolded. I had deliberately not opined on the root cause, allowing folks to reach their own conclusions. I will be very explicit here: Based on the experimental outcome, the X12S exhibited no difference in behavior from other transmitters running OpenTx 2.3.7. It was only the latest Irx4 hardware that solved the problem.

Sure, but I can show you a TMX5 that Horus X10S Express and Jumper T16 both flash perfectly, but X12S does not.

**Comment by kilrah** [comment]
My modules that don't flash in the 12S flash just fine in all other radios, so it's neither the module nor the common code, but something in the hardware-specific code for the 12S, probably an init or deinit not done right. Been pretty clear for quite a while...

**Comment by kkriegel** [comment]
@kilrah, @robustini,
@robustini's response made me go back and redo the matrix perimeter. My concern was that I did not repeat his result. In doing so, I discovered a step I missed. This is the hazard of working alone at home during a pandemic lockdown, with no others looking over your shoulder. :-)

I repeated the tests systematically on the different radios, with and without the module powered. My so-called good revision 1.3 Irx4+ upgrades just fine, with or without the module being powered before the attempt, on the Qx7 and X9D+ only. On the 2 X12S transmitters, I must start the upgrade without the module being powered.

On the X12S upgrades, with power applied before the start, the module fails to enter DFU mode after reset/repower.

**Comment by jornkaa** [comment]
Thank you very much!
The MultiBoot-Reloader worked perfectly :-)

I had firmware 1.2.0.95 on mine.  - and was because of that not able to do
the flash from the radio.
when i did the flashmulti routine after this, it all just went perfectly
well.

I have not tried flashing from the radio, but since the min req version on
the module 1.2.1.something.
when next upgrade is required i expect this to go fairly smoooth.

thanks!
Have a great day :-)

BR
- Jørn

søn. 7. jun. 2020 kl. 23:10 skrev kkriegel <notifications@github.com>:

**Comment by robustini** [comment]
Amen, if this is not possible it will flash by turning off the module first.

**Comment by robustini** [comment]
Fixed with this [commit](https://github.com/opentx/opentx/commit/af17efaddaaa75bedddf62113ef2ef2c18955407), thanks @3djc!
